File: python3/RemoveDuplicatesII.py
# Time: O(n), Space: O(n)

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if (head == None):
            return None
        
        dup = set()
        
        curNode = head
        while (curNode.next):
            if curNode.val == curNode.next.val:
                dup.add(curNode.val)
            
            curNode = curNode.next        

        if len(dup) == 0:
            return head

        dummyHead = ListNode()
        dummyHead.next = head
        
        prev = dummyHead
        curNode = head
        
        while (curNode):
            if curNode.val in dup:
                prev.next = curNode.next
            else:
                prev = curNode
            
            curNode = curNode.next
        
        return dummyHead.next

File: c++/RemoveDuplicatesII.cpp
// Time: O(n), Space: O(n)

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == NULL) {
            return head;
        }

        unordered_set<int> duplicates;
        ListNode* curNode = head;
        
        while (curNode->next) {
            if (curNode->val == curNode->next->val)
                duplicates.insert(curNode->val);
            curNode = curNode->next;
        }
        
        if (!duplicates.size()) {
            return head;
        }
        
        ListNode dummyHead;
        dummyHead.next = head;
        
        ListNode* prev = &dummyHead;
        curNode = head;
        
        while (curNode) {
            if (duplicates.find(curNode->val) != duplicates.end()) {
                prev->next = curNode->next;
            } else {
                prev = curNode;
            }
            curNode = curNode->next;
        }
        
        return dummyHead.next;
    }
};
